### DISCUSSION TITLE
Context Propagation over TCP (Or any protocol lower than HTTP in the OSI stack)

### DISCUSSION THREAD

**Discussion by ca-schue** [open-telemetry/community#1966] [Community Discussions]
As mentioned in my other discussion thread (https://github.com/open-telemetry/community/discussions/1960), I am currently looking into the possibilities and limits of using OpenTelemetry in a constrained distributed system.

As far as I could find out, the main use case for OpenTelemetry spans are protocols using HTTP which make propergating context a feasible task. 

However for my use case, OpenTelemetry would be the most helpful for providing TCP/IP observability. To pe more precise: My distributed system is more of a real-time and safety-critical cyber-physical system where most of the communication happens over protocols underneath HTTP in the OSI stack. (TCP/UDP, Socks, IP, IPSec ...)

Is OpenTelemetry still the right framework for this? How can OTLP context propagation be made possible for TCP or even IP?

**Comment by svrnm** [comment]
Hey @ca-schue, context propagation is not limited to HTTP, it's just a case where propagation is "easy". Take a look at some of the high level documents on that topic:

* https://opentelemetry.io/docs/concepts/context-propagation/
* https://opentelemetry.io/docs/specs/otel/context/api-propagators/

For some languages we even have some docs around propagation (and some hints how you can do your own):

* https://opentelemetry.io/docs/languages/erlang/propagation/
* https://opentelemetry.io/docs/languages/js/propagation/
* https://opentelemetry.io/docs/languages/ruby/
* https://opentelemetry.io/docs/languages/python/propagation/

The challenge with lower protocols may be that you need to find a place where to put the properties you need for context propagation.

**Reply by ca-schue** [reply]
Thanks for the provided documentation. I was already afraid this would boil down to stuffing tracing context in unused header fields which is almost impossible regarding TCP. What I dislike about this is therefore having to instrument something like a Socks proxy. 

Maybe encapsulating TCP/UDP or even IP packets would be the way to go for this. (e.g. [libtins](https://libtins.github.io/)) Is this within the scope of the Propagators API?
